### PR TITLE
[Enhance] Merge EditVal/TestLoop and  EditPreprocessor to `dev-1.x`

### PR DESCRIPTION
**Pull request by LeoXing1996**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Merge refactored loops and preprocessor to `dev-1.x`.

## Modification

1. EditDataSample is refactored in #1573 
2. EditDatapreprocessor is refactored in #1537 
3. Metric and Loops are refactored in #1540 
4. Support `stack` and `split` for `EditDataSample` #1669
5. Revise the forward logic for GAN models for easy outstanding #1670 

## Hightlights

1. GenValLoop` and `MultiValLoop` has been merged to `EditValLoop`, `GenTestLoop` and `MultiTestLoop` has been merged to `EditTestLoop`. Use case:
```python
    Case 1: metrics on a single dataset

    >>> # add the following lines in your config
    >>> # 1. use `EditValLoop` instead of `ValLoop` in MMEngine
    >>> val_cfg = dict(type='EditValLoop')
    >>> # 2. specific EditEvaluator instead of Evaluator in MMEngine
    >>> val_evaluator = dict(
    >>>     type='EditEvaluator',
    >>>     metrics=[
    >>>         dict(type='PSNR', crop_border=2, prefix='Set5'),
    >>>         dict(type='SSIM', crop_border=2, prefix='Set5'),
    >>>     ])
    >>> # 3. define dataloader
    >>> val_dataloader = dict(...)

    Case 2: different metrics on different datasets

    >>> # add the following lines in your config
    >>> # 1. use `EditValLoop` instead of `ValLoop` in MMEngine
    >>> val_cfg = dict(type='EditValLoop')
    >>> # 2. specific a list EditEvaluator
    >>> # do not forget to add prefix for each metric group
    >>> div2k_evaluator = dict(
    >>>     type='EditEvaluator',
    >>>     metrics=dict(type='SSIM', crop_border=2, prefix='DIV2K'))
    >>> set5_evaluator = dict(
    >>>     type='EditEvaluator',
    >>>     metrics=[
    >>>         dict(type='PSNR', crop_border=2, prefix='Set5'),
    >>>         dict(type='SSIM', crop_border=2, prefix='Set5'),
    >>>     ])
    >>> # define evaluator config
    >>> val_evaluator = [div2k_evaluator, set5_evaluator]
    >>> # 3. specific a list dataloader for each metric groups
    >>> div2k_dataloader = dict(...)
    >>> set5_dataloader = dict(...)
    >>> # define dataloader config
    >>> val_dataloader = [div2k_dataloader, set5_dataloader]
```

2. Support `stack` and `split` for `EditDataSample`, Use case:

```python
# Example for `split`
gen_sample = EditDataSample()
gen_sample.fake_img = outputs  # tensor
gen_sample.noise = noise  # tensor
gen_sample.sample_kwargs = deepcopy(sample_kwargs)  # dict
gen_sample.sample_model = sample_model  # string
# set allow_nonseq_value as True to copy non-sequential data (sample_kwargs and sample_model for this example)
batch_sample_list = gen_sample.split(allow_nonseq_value=True)  

# Example for `stack`
data_sample1 = EditDataSample()
data_sample1.set_gt_label(1)
data_sample1.set_tensor_data({'img': torch.randn(3, 4, 5)})
data_sample1.set_data({'mode': 'a'})
data_sample1.set_metainfo({
    'channel_order': 'rgb',
    'color_flag': 'color'
})
data_sample2 = EditDataSample()
data_sample2.set_gt_label(2)
data_sample2.set_tensor_data({'img': torch.randn(3, 4, 5)})
data_sample2.set_data({'mode': 'b'})
data_sample2.set_metainfo({
    'channel_order': 'rgb',
    'color_flag': 'color'
})
data_sample_merged = EditDataSample.stack([data_sample1, data_sample2])
```

## BC-breaking (Optional)

1. For BaseGAN/CondGAN models, real images are passed from `data_samples.gt_img` instead of `inputs['img']`
2. `GenValLoop` and `MultiValLoop` has been merged to `EditValLoop`, `GenTestLoop` and `MultiTestLoop` has been merged to `EditTestLoop`. 
3. `GenDataPreprocessor` has been merged into `EditDataPreprocessor`, 
    * No changes are required other than changing the `type` field in config. 
    * Users do not need to define `input_view` and `output_view` since we will infer the shape of mean automatically.
    * In evaluation stage, all tensors will be converted to `BGR` (for three-channel images) and `[0, 255]`.
4. `PixelData` has been removed

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmedit/evaluation/metrics/ms_ssim.py [hunk 1]
```diff
@@ -291,20 +291,23 @@ def process(self, data_batch: dict, data_samples: Sequence[dict]) -> None:
...
                 # get img tensor
-                fake_img_ = fake_img_['fake_img']['data']
+                fake_img_ = fake_img_['fake_img']
             fake_imgs.append(fake_img_)
         minibatch = torch.stack(fake_imgs, dim=0)
 
         assert minibatch.shape[0] % 2 == 0, 'batch size must be divided by 2.'
-        minibatch = ((minibatch + 1) / 2)
-        minibatch = minibatch.clamp_(0, 1)
+        # minibatch = ((minibatch + 1) / 2)
```
Comment (zengyh1900) [inline mmedit/evaluation/metrics/ms_ssim.py:302]: remove these comments?

File: mmedit/models/base_models/base_conditional_gan.py [hunk 2]
```diff
@@ -176,14 +176,14 @@ def forward(self,
...
-            num_batches = get_valid_num_batches(inputs)
+            num_batches = get_valid_num_batches(inputs, data_samples)
             noise = self.noise_fn(noise, num_batches=num_batches)
             sample_kwargs = inputs.get('sample_kwargs', dict())
         num_batches = noise.shape[0]
 
         labels = self.data_sample_to_label(data_samples)
         if labels is None:
-            num_batches = get_valid_num_batches(inputs)
+            # num_batches = get_valid_num_batches(inputs)
```
Comment (zengyh1900) [inline mmedit/models/base_models/base_conditional_gan.py:186]: remove unused codes

File: mmedit/models/editors/cyclegan/cyclegan.py [hunk 3]
```diff
@@ -282,10 +282,10 @@ def val_step(self, data: dict) -> SampleList:
...
             for src_domain in self._reachable_domains:
                 target_domain = self.get_other_domains(src_domain)[0]
-                setattr(gen_sample, f'gt_{src_domain}',
-                        PixelData(data=inputs_dict[f'img_{src_domain}'][idx]))
-                setattr(gen_sample, f'fake_{src_domain}',
-                        PixelData(data=outputs[f'img_{src_domain}'][idx]))
+                fake_img = outputs[f'img_{target_domain}'][idx]
+                fake_img = self.data_preprocessor.destruct(
+                    fake_img, data_samples[idx], f'img_{target_domain}')
+                setattr(gen_sample, f'fake_{target_domain}', fake_img)
```
Comment (zengyh1900) [inline mmedit/models/editors/cyclegan/cyclegan.py:288]: I think using `setattr` to add attributes in `gen_sample` can be a bit hacky. 
I suggest using 
`gen_sample.set_field(fake_img, f'fake_{target_domain}', dtype=tensor)`
or 
`gen_sample.set_tensor_data(f'fake_{target_domain}', fake_img)`

File: mmedit/structures/edit_data_sample.py [hunk 4]
```diff
@@ -112,542 +120,91 @@ class EditDataSample(BaseDataElement):
...
+            data (dict): A dict contains annotations of image or
+                model predictions.
+        """
+        assert isinstance(data,
+                          dict), f'data should be a `dict` but got {data}'
+        for k, v in data.items():
+            if k == 'gt_label':
+                self.set_gt_label(v)
+            else:
+                setattr(self, k, all_to_tensor(v))
```
Comment (zengyh1900) [inline mmedit/structures/edit_data_sample.py:207]:
```suggestion
                self.set_field(all_to_tensor(v), k, dtype=torch.tensor)
```

File: mmedit/models/editors/eg3d/eg3d.py [hunk 5]
```diff
@@ -146,11 +146,12 @@ def pack_to_data_sample(self,
             assert isinstance(v, torch.Tensor), (
                 f'Output must be tensor. But \'{k}\' is type of '
                 f'\'{type(v)}\'.')
-            if v.ndim == 4 and v.shape[1] == 3:
-                setattr(data_sample, k, PixelData(data=v[index]))
-            else:
-                # NOTE: hard code here, we assume all tensor are [bz, ...]
-                setattr(data_sample, k, v[index])
+            # if v.ndim == 4 and v.shape[1] == 3:
```
Comment (zengyh1900) [inline mmedit/models/editors/eg3d/eg3d.py:149]: remove unused comments

File: mmedit/models/editors/ggan/ggan.py [hunk 6]
```diff
@@ -79,7 +79,7 @@ def train_discriminator(self, inputs: dict,
         Returns:
             Dict[str, Tensor]: A ``dict`` of tensor for logging.
         """
-        real_imgs = inputs['img']
+        real_imgs = torch.stack([data.gt_img for data in data_samples])
```
Comment (zengyh1900) [inline mmedit/models/editors/ggan/ggan.py:82]: why do we still stack tensor in models' forward?
Have we decided to stack tensors in `data_preprocessor`?

Comment (zengyh1900) [reply mmedit/models/editors/ggan/ggan.py:82]: We may need check this point for all models

File: mmedit/models/editors/pggan/pggan.py [hunk 7]
```diff
@@ -161,13 +161,15 @@ def forward(self,
...
+        outputs = self.data_preprocessor.destruct(outputs, data_samples)
 
         if sample_model == 'ema/orig':
             _model = self.generator
             outputs_orig = _model(
                 noise,
                 curr_scale=curr_scale,
                 transition_weight=transition_weight)
+            outputs = self.data_preprocessor.destruct(outputs, data_samples)
             outputs = dict(ema=outputs, orig=outputs_orig)
```
Comment (zengyh1900) [inline mmedit/models/editors/pggan/pggan.py:173]: here, `outputs` is constructed as a dictionary, and in the following codes, we fetch the values and pass them to `gen_sample`. So why not directly pass the values to the `gen_sample`. Besides, why naming the data_sample as `gen_sample`?

Comment (zengyh1900) [reply mmedit/models/editors/pggan/pggan.py:173]: check these points for all the models.

File: mmedit/models/editors/pix2pix/pix2pix.py [hunk 8]
```diff
@@ -188,12 +188,14 @@ def test_step(self, data: dict) -> SampleList:
             gen_sample = EditDataSample()
             if data_samples:
                 gen_sample.update(data_samples[idx])
-            setattr(gen_sample, f'gt_{target_domain}',
-                    PixelData(data=inputs_dict[f'img_{target_domain}'][idx]))
-            setattr(gen_sample, f'fake_{target_domain}',
-                    PixelData(data=outputs['target'][idx]))
-            setattr(gen_sample, f'gt_{source_domain}',
-                    PixelData(data=inputs_dict[f'img_{source_domain}'][idx]))
+            # setattr(gen_sample, f'gt_{target_domain}',
```
Comment (zengyh1900) [inline mmedit/models/editors/pix2pix/pix2pix.py:191]: Clear unused comments in this PR, pls.

File: mmedit/visualization/gen_visualizer.py [hunk 9]
```diff
@@ -1,6 +1,5 @@
 # Copyright (c) OpenMMLab. All rights reserved.
```
Comment (zengyh1900) [inline mmedit/visualization/gen_visualizer.py:1]: `gen_visualizer` is confusing. Reconsider this point.

**Review by zengyh1900 — CHANGES_REQUESTED** [review]
This PR involves significant changes. 
Please list as detailed as possible use_case, migration notes, etc.

**Comment by zengyh1900** [comment]
Hi @LeoXing1996 feel free to discuss any comments. If I misunderstand some usage, please correct me.
